### ISSUE TITLE
Lower framerates / performance

### ISSUE THREAD

**Issue by carlobrok** [Qengineering/MobileNet_SSD_OpenCV_TensorFlow#1]
I tested your program on my Raspberry Pi, but I get much lower framerates.
Your source code has been adjusted a little so that I can see the images over the network. But I don't lose performance there (already tested).

system information:
* Raspberry Pi 4B 2GB edition
* Raspbian lite buster 
* OpenCV 4.3.0

I can achieve these framerates:
ssd mobilenet v1 : 1.78 FPS
ssdlite mobilenet v2 : 2.4 FPS

Only 140MB RAM is used. So using the 4GB edition wouldn't improve the performance? Or am I wrong?

Which operating system and hardware do you use?

Thanks!

**Comment by carlobrok** [comment]
I have manually compiled the code using g++ and optimization level 3 (-O3). I have not used Code::Blocks.

**Comment by Qengineering** [comment]
I'm sorry to hear about your disappointment. There must be some other process holding your RPi occupied. I've tested it just again, and still getting the 3.2 FPS. Please check if your RPi is running at 85% processing and no other applications are running at the background. See the _htop_ screendump below of the processes running on my RPi.
![htop](https://user-images.githubusercontent.com/44409029/80203479-8287b600-8627-11ea-985a-868209330e4f.png)
.

**Comment by carlobrok** [comment]
There are no other significant processes running. Only the root processes started by raspbian. Does raspbian run on your pi too? In [this](https://github.com/Qengineering/TensorFlow_Lite_RPi_64-bits) repo you're using Ubuntu. Is your Pi overclocked?

This is my screendump. The program uses even more % of the cpu:

![screendump](https://user-images.githubusercontent.com/44724164/80205908-be247f00-862b-11ea-90f4-f226cf110c37.png)

**Comment by Qengineering** [comment]
Dear Carlo,
I'm using the regular Raspbian os (Linux raspberrypi 4.19.97-v7l+ #1294 SMP Thu Jan 30 13:21:14 GMT 2020 armv7l GNU/Linux) installed by NOOBS 3.3.1. No overclocking here, just plane 1500 MHz.
The only reason I can think of for now is the installation of OpenCV. As if not all the internal CPU registers are used or something like that. Given your high CPU usage and low throughput. 
Below my OpenCV 4.3.0 build info used in the examples:

pi@raspberrypi:~ $ python
Python 2.7.16 (default, Oct 10 2019, 22:02:15) 
[GCC 8.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import cv2
>>> cv2.__version__
'4.3.0'
>>> print cv2.getBuildInformation()

General configuration for OpenCV 4.3.0 =====================================
  Version control:               unknown

  Extra modules:
    Location (extra):            /home/pi/opencv_contrib/modules
    Version control (extra):     unknown

  Platform:
    Timestamp:                   2020-04-22T13:14:18Z
    Host:                        Linux 4.19.97-v7l+ armv7l
    CMake:                       3.13.4
    CMake generator:             Unix Makefiles
    CMake build tool:            /usr/bin/make
    Configuration:               RELEASE

  CPU/HW features:
    Baseline:                    VFPV3 NEON
      requested:                 DETECT
      required:                  VFPV3 NEON

  C/C++:
    Built as dynamic libs?:      YES
    C++ standard:                11
    C++ Compiler:                /usr/bin/c++  (ver 8.3.0)
    C++ flags (Release):         -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -mfpu=neon -fvisibility=hidden -fvisibility-inlines-hidden -fopenmp -O3 -DNDEBUG  -DNDEBUG
    C++ flags (Debug):           -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -mfpu=neon -fvisibility=hidden -fvisibility-inlines-hidden -fopenmp -g  -O0 -DDEBUG -D_DEBUG
    C Compiler:                  /usr/bin/cc
    C flags (Release):           -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Winit-self -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -mfpu=neon -fvisibility=hidden -fopenmp -O3 -DNDEBUG  -DNDEBUG
    C flags (Debug):             -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Winit-self -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -mfpu=neon -fvisibility=hidden -fopenmp -g  -O0 -DDEBUG -D_DEBUG
    Linker flags (Release):      -Wl,--gc-sections -Wl,--as-needed  
    Linker flags (Debug):        -Wl,--gc-sections -Wl,--as-needed  
    ccache:                      NO
    Precompiled headers:         NO
    Extra dependencies:          dl m pthread rt
    3rdparty dependencies:

  OpenCV modules:
    To be built:                 aruco bgsegm bioinspired calib3d ccalib core datasets dnn dnn_objdetect dnn_superres dpm face features2d flann freetype fuzzy gapi hdf hfs highgui img_hash imgcodecs imgproc intensity_transform line_descriptor ml objdetect optflow phase_unwrapping photo plot python2 python3 quality rapid reg rgbd saliency shape stereo stitching structured_light superres surface_matching text tracking ts video videoio videostab xfeatures2d ximgproc xobjdetect xphoto
    Disabled:                    world
    Disabled by dependency:      -
    Unavailable:                 alphamat cnn_3dobj cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev cvv java js matlab ovis sfm viz
    Applications:                perf_tests apps
    Documentation:               NO
    Non-free algorithms:         YES

  GUI: 
    GTK+:                        YES (ver 3.24.5)
      GThread :                  YES (ver 2.58.3)
      GtkGlExt:                  NO

  Media I/O: 
    ZLib:                        /usr/lib/arm-linux-gnueabihf/libz.so (ver 1.2.11)
    JPEG:                        /usr/lib/arm-linux-gnueabihf/libjpeg.so (ver 62)
    WEBP:                        build (ver encoder: 0x020f)
    PNG:                         /usr/lib/arm-linux-gnueabihf/libpng.so (ver 1.6.36)
    TIFF:                        build (ver 42 - 4.0.10)
    JPEG 2000:                   /usr/lib/arm-linux-gnueabihf/libjasper.so (ver 1.900.1)
    OpenEXR:                     build (ver 2.3.0)
    HDR:                         YES
    SUNRASTER:                   YES
    PXM:                         YES
    PFM:                         YES

  Video I/O:
    DC1394:                      YES (2.2.5)
    FFMPEG:                      YES
      avcodec:                   YES (58.35.100)
      avformat:                  YES (58.20.100)
      avutil:                    YES (56.22.100)
      swscale:                   YES (5.3.100)
      avresample:                NO
    GStreamer:                   NO
    v4l/v4l2:                    YES (linux/videodev2.h)

  Parallel framework:            TBB (ver 2020.1 interface 11101)

  Trace:                         YES (with Intel ITT)

  Other third-party libraries:
    Lapack:                      NO
    Custom HAL:                  YES (carotene (ver 0.0.1))
    Protobuf:                    build (3.5.1)

  OpenCL:                        YES (no extra features)
    Include path:                /home/pi/opencv/3rdparty/include/opencl/1.2
    Link libraries:              Dynamic load

  Python 2:
    Interpreter:                 /usr/bin/python2.7 (ver 2.7.16)
    Libraries:                   /usr/lib/arm-linux-gnueabihf/libpython2.7.so (ver 2.7.16)
    numpy:                       /usr/lib/python2.7/dist-packages/numpy/core/include (ver 1.16.2)
    install path:                lib/python2.7/dist-packages/cv2/python-2.7

  Python 3:
    Interpreter:                 /usr/bin/python3 (ver 3.7.3)
    Libraries:                   /usr/lib/arm-linux-gnueabihf/libpython3.7m.so (ver 3.7.3)
    numpy:                       /usr/lib/python3/dist-packages/numpy/core/include (ver 1.16.2)
    install path:                lib/python3.7/dist-packages/cv2/python-3.7

  Python (for build):            /usr/bin/python2.7

  Java:                          
    ant:                         NO
    JNI:                         NO
    Java wrappers:               NO
    Java tests:                  NO

  Install to:                    /usr/local
-----------------------------------------------------------------

**Comment by carlobrok** [comment]
I appreciate your effort and will compare it with my OpenCV build configuration!

**Comment by carlobrok** [comment]
The main differences are here:

Configuration |  My OpenCV build  |    Yours 
--------------|-------------|------------------------
Version | 4.3.0-dev | 4.3.0 
**VFPV3 NEON** |  **Baseline: <br> requested: DETECT <br> disabled: VFPV3 NEON** | **Baseline: VFPV3 NEON <br> requested: DETECT <br> required: VFPV3 NEON**
Non-free algorithms | NO | YES
GUI | QT: YES <br> QT OpenGL supports: YES <br> OpenGL support: YES <br> GTK+ & VTK: NO | GTK+: YES <br> GThread : YES
**Parallel framework** | **pthreads** | **TBB**
Custom HAL | NO |  YES
Python | python3 | python2.7


I think python is irrelevant since I am using c++, also GUI is irrelevant because there is no performance loss displaying the image. 
As [this page](https://www.pyimagesearch.com/2017/10/09/optimizing-opencv-on-the-raspberry-pi/) suggests, major improvements in performance can be made with NEON, some with TBB.. I also read that a 64 bit operating system like Ubuntu server can make the programs run faster. I'll test these things.

**Comment by Qengineering** [comment]
There you have your answer! NEON and VFPV3 are crucial when it comes to the speed of the convolutions in your neural network. TBB can also add some additional speed. Please look at our page: [https://qengineering.eu/install-opencv-4.3-on-raspberry-pi-4.html](url) for the correct options on your CMAKE build of OpenCV.

**Comment by carlobrok** [comment]
Great! I got ~3,26 FPS on mobilenet V1
